app/services/search_indexer.rb

```diff
@@ -64,7 +71,16 @@ def self.update_index(table:, id:, a_weight: nil, b_weight: nil, c_weight: nil,
         array
       end
 
-    tsvector = "#{tsvector} #{additional_lexemes.join(" ")}"
+    extra_domain_word_terms =
+      if additional_words.length > 0
+        # Note: postions are off, keeping them 100 correct is very complex
+        # cause some words may stem back to nothing due to being stop words
+        # we do not have any <=> searches so the possible pollution here is not
+        # too big
+        DB.query_single("SELECT to_tsvector(?, ?)", stemmer, additional_words.uniq.join(" ")).first
```

Comment:
This makes me uncomfortable considering that we're not just losing position but also the weights. 

> keeping them 100 correct is very complex cause some words may stem back to nothing due to being stop words
        
I don't think `to_tsvector` remove stop words at all.

```
discourse_development=# SELECT TO_TSVECTOR('simple', 'a b c d e');
          to_tsvector          
-------------------------------
 'a':1 'b':2 'c':3 'd':4 'e':5
```

Therefore, I think it is quite possible for us to retain the position and weight in the tsvector.

Reply:
```
    tsvector
      .scan(/'(([a-zA-Z0-9]+\.)+[a-zA-Z0-9]+)'\:([\w+,]+)/)
      .reduce(additional_lexemes) do |array, (lexeme, _, positions)|
        count = 0

        if lexeme !~ /\A(\d+\.)?(\d+\.)*(\*|\d+)\z/
          loop do
            count += 1
            break if count >= 10 # Safeguard here to prevent infinite loop when a term has many dots
            term, _, remaining = lexeme.partition(".")
            break if remaining.blank?

            additional_words << [term, positions]

            array << "'#{remaining}':#{positions}"
            lexeme = remaining
          end
        end

        array
      end

    extra_domain_word_terms =
      if additional_words.length > 0
        lexemes =
          DB.query_single(
            "SELECT STRIP(TO_TSVECTOR(?, ?))",
            stemmer,
            additional_words.transpose[0].join(" "),
          ).first

        lexemes
          .split(" ")
          .map
          .with_index { |lexeme, index| "#{lexeme}:#{additional_words[index][1]}" }
          .join(" ")
      end
```

We can do something like this to retain both position and weight.

Reply:
You tested with a simple stemmer, try with English , I guess we can use this approach, but we do bloat the index a bit this way

Reply:
> You tested with a simple stemmer, try with English

Ahh icic. In that case, I think we can just execute the `TO_TSVECTOR` function once per word but in a single query? Quick solution spike below: 

```
SELECT 
  CONCAT(STRIP(TO_TSVECTOR(<stemmer>, 'word1')), ':<position>'),
  CONCAT(STRIP(TO_TSVECTOR(<stemmer>, 'word2')), ':<position>'),
```

Reply:
that is one monster query ... you could be selecting 100s of columns :cry: but yeah it can work

```
discourse_development=# select * from to_tsvector('english', 'hellos the houses the hello');
     to_tsvector      
----------------------
 'hello':1,5 'hous':3
(1 row)
```


Reverse mapping :arrow_double_up: though is annoying albeit technically possible, you treat positions as offsets into the original array. positions are maintained.